src/webServer/webServer.ts

```diff
@@ -0,0 +1,193 @@
+        }
+
+        hasLevel(level: LogLevel): boolean {
+            return this.level >= level;
+        }
+
+        loggingEnabled(): boolean {
+            return true;
+        }
+
+        perftrc(s: string): void {
+            this.msg(s, Msg.Perf);
+        }
+
+        info(s: string): void {
+            this.msg(s, Msg.Info);
+        }
+
+        err(s: string) {
+            this.msg(s, Msg.Err);
+        }
+
+        startGroup(): void {
+            ++this.currentGroupCount;
+        }
+
+        endGroup(): void {
+            this.currentGroupCount = Math.max(0, this.currentGroupCount - 1);
+        }
+
+        msg(s: string, type: Msg = Msg.Err): void {
+            s = `${type} ${this.seq.toString()} [${nowString()}] ${s}`;
+
+            switch (type) {
+                case Msg.Info:
+                    this.write("info", s);
+                    break;
+
+                case Msg.Perf:
+                    this.write("perf", s);
+                    break;
+
+                case Msg.Err:
+                default:
+                    this.write("error", s);
+                    break;
+            }
+
+            if (this.currentGroupCount === 0) {
+                this.seq++;
+            }
+        }
+
+        getLogFileName(): string | undefined {
+            return undefined;
+        }
+
+        private write(level: MessageLogLevel, body: string) {
+            this.host.writeMessage(<LoggingMessage>{
+                type: "log",
+                level,
+                body,
+            });
+        }
+    }
+
+    export function createWebSystem(host: WebHost, args: string[], getExecutingFilePath: () => string): ServerHost {
+        const returnEmptyString = () => "";
+        const getExecutingDirectoryPath = memoize(() => memoize(() => ensureTrailingDirectorySeparator(getDirectoryPath(getExecutingFilePath()))));
+        // Later we could map ^memfs:/ to do something special if we want to enable more functionality like module resolution or something like that
+        const getWebPath = (path: string) => startsWith(path, directorySeparator) ? path.replace(directorySeparator, getExecutingDirectoryPath()) : undefined;
+        return {
+            args,
+            newLine: "\r\n", // This can be configured by clients
+            useCaseSensitiveFileNames: false, // Use false as the default on web since that is the safest option
+            readFile: path => {
+                const webPath = getWebPath(path);
+                return webPath && host.readFile(webPath);
+            },
+
+            write: host.writeMessage.bind(host),
+            watchFile: returnNoopFileWatcher,
+            watchDirectory: returnNoopFileWatcher,
+
+            getExecutingFilePath: () => directorySeparator,
+            getCurrentDirectory: returnEmptyString, // For inferred project root if projectRoot path is not set, normalizing the paths

```

Comment:
I bet the difference between `""` and `undefined` is important, but I can't tell how.

Reply:
currentDirectory is suppose to be string and so it needs to return empty string.. we directory concatenate those strings so undefined would need too much checks i guess that was logic behind that.